### PR TITLE
Feature/unity react testing

### PR DESCRIPTION
**Pull request by RonakReyhani**

### CODE REVIEW

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+
+During the execution of this command in the Unity world, React will actively listen for the event that is returned from Unity. Once the event is received, we can capture the player's position and save the screenshot file accordingly.
+
+### React Layer
+
+To begin, let's define the Unity Event Type and the necessary commands that we will execute.
+
+In the integration between Unity and React, we need to configure interfaces and Unity event types that will facilitate communication between the two frameworks. This Event Type will serve as a structured format for exchanging information and triggering actions.
+
+Additionally, we must identify the specific commands that React will execute to interact with Unity. These commands will be responsible for instructing Unity to perform tasks such as retrieving the player's position and capturing a screenshot of their view.
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:29]: `player's` -> `Player's`

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+
+By defining the Unity Event Type and the required commands, we lay the foundation for seamless communication and collaboration between Unity and React in our integration testing scenario.
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:31]: This line can be deleted:

> By defining the Unity Event Type and the required commands, we lay the foundation for seamless communication and collaboration between Unity and React in our integration testing scenario.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+`deps`: An array of dependencies that the effect depends on, similar to the dependencies array in the useEffect hook.
+`eventName`: The name of the Unity event to subscribe to.
+
+To make the Unity context available throughout our entire app, we simply need to wrap our app with the `UnityContextProvider` that we created. By doing this, all components within the app will have access to the Unity events and functionalities. Here's how you can wrap your app with the `UnityContextProvider`:
+
+```tsx
+// app.tsx
+
+import { UnityContextProvider } from "./UnityContextProvider";
+import { CssBaseline } from "@mui/material";
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:215]: @MUI refences can be removed, remove the noise.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+
+import React, { FC } from "react";
+import GlobalEventProvider from "../components/GlobalEventProvider";
+
+const App: FC<AppProps> = ({ Component, pageProps }) => {
+  return (
+    <UnityContextProvider>
+      <CssBaseline />
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:223]: remove

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+To make the Unity context available throughout our entire app, we simply need to wrap our app with the `UnityContextProvider` that we created. By doing this, all components within the app will have access to the Unity events and functionalities. Here's how you can wrap your app with the `UnityContextProvider`:
+
+```tsx
+// app.tsx
+
+import { UnityContextProvider } from "./UnityContextProvider";
+import { CssBaseline } from "@mui/material";
+
+import React, { FC } from "react";
+import GlobalEventProvider from "../components/GlobalEventProvider";
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:218]: remove

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+};
+
+const AppWrapper: FC<AppProps> = (appProps) => {
+  return <App {...appProps} />;
+};
+
+export default AppWrapper;
+```
+
+Now, the Unity context will be accessible throughout the app, allowing you to subscribe to Unity events, interact with Unity functionalities, and utilize custom hooks such as useUnityEventSubscription and useUnityEventEffect in any component within your app.
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:236]: "useUnityEventSubscription" -> "`useUnityEventSubscription`"

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+
+If you are new to the BBD (Behavior-Driven Development) concept, I recommend reading the article titled ["Example Mapping in Practice"](https://blog.mechanicalrock.io/2023/03/21/example-mapping-in-practice.html#:~:text=Example%20Mapping%20is%20a%20technique,common%20understanding%20amongst%20the%20team) This article provides a comprehensive overview of BDD, its principles, and how it can be applied in software development.
+
+### Expected Behavior
+
+Now that we have completed all the necessary setup and configuration, it's time to start developing our first component using Behavioral Development Driven (BDD) principles. BDD helps us define the expected behavior of our component in a clear and understandable manner. Here's the expected behavior for our component:
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:248]: the words "Behavioral Development Driven" has already been defined earler, from this point onwards you can just specify "BDD"

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+
+```tsx
+// addScene.feature
+Scenario: Add New Scene
+  Given Drew views the animation details
+  When Drew adds a new scene
+  Then The scene has to be added
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:255]: "Then The scene has to be added" -> "Then the new scene should be added to the open animation"

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+  test("Add New Scene", ({ given, when, and, then }) => {
+    given("Drew views the animation details", async () => {
+      // wait to see the animation details
+      getAnimationView();
+    });
+    when("Drew adds a new scene", async () => {
+      // Add new scene that will trigger a unity event
+      await addNewScene();
+    });
+    then("The scene has to be added", async () => {
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:296]: Update then() string with the revised name

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+async function getPlayerPositionUnityEvent(x: number, y: number, z: number) {
+  const event = JSON.stringify({
+    eventName: "receivedAnimationSceneDetails",
+    args: JSON.stringify({
+      playerPosition: { x: x, y: y, z: z },
+      screenshotData: "base64 image data",
+    }),
+  });
+  act(() => {
+    const handler = eventHandler(true, "");
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:334]: not sure why `true` is the first argument and `empty string` is the second value. This needs to be explained.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+
+export async function addNewScene(index?: number) {
+  await waitFor(() => {
+    expect(screen.queryByTestId("button-add-scene")).not.toBeNull();
+  });
+  act(() => {
+    fireEvent.click(screen.queryByTestId("button-add-scene")!);
+  });
+
+  await waitFor(() => {
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:350]: this waitFor() is not required, if you declare it this way the inners will be executed roughly every 100ms, which we really don't want to happen.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+    // this would be our helper function to mock the unity response
+    // so now, each time we add a new scene the mocked player position is return and
+    // React app will simply update the app state with the returned data.
+    getPlayerPositionUnityEvent(50, 50, 120);
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:354]: this should be prefixed with `await `

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+
+function SceneWrapper(scene: SceneDetails, index: number, scenes: SceneDetails[]) {
+  return (
+    <React.Fragment key={scene?.id || index}>
+      <SceneCard index={index} scene={scene} nextScene={scenes[index + 1]} />
+    </React.Fragment>
+  );
+}
+
+export default ViewAnimation;
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:395]: mmm... wayyyyyy too much elaborated code example this blog doesn't need, the example should be clear and concise.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+const wrapper: FC<{}> = ({ children }) => (
+  <UnityContextProvider>
+    <AnimationProvider>{children}</AnimationProvider>
+  </UnityContextProvider>
+);
+
+describe("Should Call getAnimationSceneDetails command", () => {
+  it("should execute the getSceneDetails command when clicked add scene", async () => {
+    const executeCommandFn = jest.fn();
+    // @ts-ignore
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:530]: remove this comment, it's our old-tsconfig specificness which is why this is required, afresh project might not require this.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+});
+```
+
+### Conclusion
+
+In conclusion, integrating Unity with React opens up exciting possibilities for building immersive and interactive applications. By adopting the Behavior-Driven Development (BDD) approach, we can ensure that our software meets the desired behavior and aligns with business objectives. BDD emphasizes collaboration, communication, and test automation, enabling teams to work together effectively and deliver high-quality applications.
+
+Through the use of tools like Jest Cucumber and React Testing Library, we can write readable and expressive test scenarios that capture the behavior of our Unity and React components. These tests serve as living documentation, providing clarity on the expected behavior and enabling easy maintenance and collaboration among team members.
+
+If you're eager to explore more about the example codes provided in this post, you can check out the accompanying GitHub repository. You can find the GitHub repository [here](https://github.com/MechanicalRock/unity-react-example).
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:556]: It's nice you've done this, but we're trying not to give people the answers to a lot of this Unity and React development because we want to win the jobs to help implement these changes for clients. @Hamo69  to have the final yes/no say on showing the solution. Most of this comment applies to the blog; requires Randy.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+
+    await waitFor(() => {
+      expect(screen.getByTestId("button-add-scene")).not.toBeNull();
+    });
+    act(() => {
+      fireEvent.click(screen.getByTestId("button-add-scene"));
+    });
+    await waitFor(() => {
+      expect(executeCommandFn).toHaveBeenCalledWith({
+        args: { height: 180, width: 134 },
```
Comment (qt-pixels) [inline _posts/2023-07-12-React-unity-BDD.markdown:542]: where did these values come from? I can't see them in this test.

**Review by qt-pixels — CHANGES_REQUESTED** [review]

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 17]
```diff
@@ -0,0 +1,560 @@
+---
+layout: postv2
+font: serif
+title: "Unity & React Behavioral Development Driven"
```
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:4]: Behaviour not Behavioural

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+export default AppWrapper;
+```
+
+Now, the Unity context will be accessible throughout the app, allowing you to subscribe to Unity events, interact with Unity functionalities, and utilize custom hooks such as useUnityEventSubscription and useUnityEventEffect in any component within your app.
+
+By wrapping your app with the UnityContextProvider, you establish a bridge between Unity and your React app, enabling seamless integration and communication between the two environments.
+
+### BDD Testing & Unity Event Mocks
+
+BDD (Behavior-Driven Development) is an agile software development methodology that focuses on collaboration and communication between developers, quality assurance (QA) engineers, and business stakeholders. It aims to ensure that software development efforts align with business requirements and produce valuable outcomes. In BDD, the behavior of the system is described in a common language that is easily understandable by all stakeholders.
```
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:242]: behaviour

File: _posts/2023-07-12-React-unity-BDD.markdown (hunk 7 above)
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:248]: Behaviour

File: _posts/2023-07-12-React-unity-BDD.markdown (hunk 9 above)
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:296]: then("The scene SHOULD be added", async () => {

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 19]
```diff
@@ -0,0 +1,560 @@
...
+// addScene.feature
+Scenario: Add New Scene
+  Given Drew views the animation details
+  When Drew adds a new scene
+  Then The scene has to be added
+```
+
+#### BDD Feature Steps
+
+Let's dive into the development of our component, considering the expected behavior outlined above. By using BDD, we can create clear and comprehensive test cases and scenarios that cover various use cases and edge cases. This ensures that our component functions as expected in different scenarios and meets the desired behavior defined through BDD. To achieve this objective, we leverage the capabilities of Jest Cucumber for loading the feature file and utilize the React Testing Library.
```
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:260]: I would use the word Examples rather than scenarios and test cases. 

By using BDD, we utilise the power of examples to create clear and comprehensive tests that cover many scenarios. This ensures that our component functions as expected in different scenarios and meets the desired behaviour. To achieve this objective, we leverage the capabilities of Jest Cucumber for loading the feature file and utilise the React Testing Library.

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 20]
```diff
@@ -0,0 +1,560 @@
...
+}));
+
+const wrapper: FC<{}> = ({ children }) => (
+  <UnityContextProvider>
+    <AnimationProvider>{children}</AnimationProvider>
+  </UnityContextProvider>
+);
+
+describe("Should Call getAnimationSceneDetails command", () => {
+  it("should execute the getSceneDetails command when clicked add scene", async () => {
```
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:528]: This is an imperative definition rather than declarative.
 What I mean by this is "when clicked add scene" assumes a certain implementation and is not behaviour specific

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 21]
```diff
@@ -0,0 +1,560 @@
...
+        name: "getAnimationSceneDetails",
+      });
+    });
+  });
+});
+```
+
+### Conclusion
+
+In conclusion, integrating Unity with React opens up exciting possibilities for building immersive and interactive applications. By adopting the Behavior-Driven Development (BDD) approach, we can ensure that our software meets the desired behavior and aligns with business objectives. BDD emphasizes collaboration, communication, and test automation, enabling teams to work together effectively and deliver high-quality applications.
```
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:552]: You need to tidy up how you refer to BDD. You have various methods. Here you use Behavior-Driven Development (BDD). I prefer English spelling without hyphens - just make it consistent. "Behaviour Driven Development"

File: _posts/2023-07-12-React-unity-BDD.markdown (hunk 8 above)
Comment (Hamo69) [reply _posts/2023-07-12-React-unity-BDD.markdown:255]: Then the scene SHOULD be added

File: _posts/2023-07-12-React-unity-BDD.markdown [hunk 22]
```diff
@@ -0,0 +1,560 @@
...
+
+If you are new to the BBD (Behavior-Driven Development) concept, I recommend reading the article titled ["Example Mapping in Practice"](https://blog.mechanicalrock.io/2023/03/21/example-mapping-in-practice.html#:~:text=Example%20Mapping%20is%20a%20technique,common%20understanding%20amongst%20the%20team) This article provides a comprehensive overview of BDD, its principles, and how it can be applied in software development.
+
+### Expected Behavior
+
+Now that we have completed all the necessary setup and configuration, it's time to start developing our first component using Behavioral Development Driven (BDD) principles. BDD helps us define the expected behavior of our component in a clear and understandable manner. Here's the expected behavior for our component:
+
+```tsx
+// addScene.feature
+Scenario: Add New Scene
```
Comment (Hamo69) [inline _posts/2023-07-12-React-unity-BDD.markdown:252]: Scenario: The one where Drew Adds a New Scene

**Review by Hamo69 — CHANGES_REQUESTED** [review]

File: _posts/2023-07-12-React-unity-BDD.markdown (hunk 20 above)
Comment (RonakReyhani) [reply _posts/2023-07-12-React-unity-BDD.markdown:528]: Hamo this is the second approach which is more of a unit test not BDD scenario. I have specified it in the explanations before the test.
